### ISSUE TITLE
Sanitizer support doesn't properly check for compiler/linker capabilities

### ISSUE THREAD

**Issue by ericonr** [mesonbuild/meson#8283]
**Describe the bug**

Meson doesn't check if the `b_sanitize` values passed to it are sane for the compiler it's currently using, which ends up with failures in the build step instead of during the configuration step. This is specially visible on musl libc systems, where, at least for now, there is no sanitizer support.

**To Reproduce**

Simply run `CC=gcc ./run_unittests.py` on a musl system, it will error out with the following message:

```
[1/19] Precompiling header '../test cases/common/13 pch/c/pch/prog.h'
[2/19] Generating makeheader with a custom command
[3/19] Compiling C object c/prog.p/prog.c.o
[4/19] Linking target c/prog
FAILED: c/prog 
gcc  -o c/prog c/prog.p/prog.c.o -fsanitize=address -Wl,--as-needed -Wl,--allow-shlib-undefined
/usr/bin/ld: cannot find libasan_preinit.o: No such file or directory
/usr/bin/ld: cannot find -lasan
collect2: error: ld returned 1 exit status
```

From my understanding of the decorator around the tests that use sanitizers, they should be skipped when unsupported, but this doesn't happen.

More examples:

- With GCC 10, which here seems to not know about the `memory` sanitizer, it will also error out (during the build) if I configure any C project with `-Db_sanitize=memory`:

```
samu: job failed: cc -Ilibpurrlib.a.p -I. -I../../home/ericonr/mypro/purr-c -fdiagnostics-color=always -fsanitize=memory -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -g '-DGETTEXT_PACKAGE="purr-c"' '-DGETTEXT_DIR="/usr/local/share/locale"' -DUSE_LIBINTL -DHAVE_PROG_INVOCATION -DHAVE_SOCK_CLOEXEC_H -DHAVE_PIPE2 -fPIC -MD -MQ libpurrlib.a.p/encrypt.c.o -MF libpurrlib.a.p/encrypt.c.o.d -o libpurrlib.a.p/encrypt.c.o -c ../../home/ericonr/mypro/purr-c/encrypt.c
cc: error: unrecognized argument to '-fsanitize=' option: 'memory'
```

- With GCC 10, which does know about the `address` sanitizer, it will instead error out during linking:

```
[16/18] Linking target tests
samu: job failed: cc  -o tests tests.p/tests.c.o -fsanitize=address -Wl,--as-needed -Wl,--no-undefined -Wl,--start-group libpurrlib.a -lbearssl -Wl,--end-group
/bin/ld: cannot find libasan_preinit.o: Arquivo ou diretório inexistente
/bin/ld: cannot find -lasan
collect2: error: ld returned 1 exit status
samu: subcommand failed
```

**Expected behavior**

Meson should error out during the configuration step, claiming that the selected sanitizer options aren't supported by the compiler or that the sanitizer libraries aren't available for linking.

**system parameters**

Native build, on Void Linux musl, Python 3.9.0, Meson 0.56.2 

This issue, once solved, can likely solve #6486 and #6487.

**Comment by dcbaker** [comment]
This is a well known problem (at least I know about it :D ). We need to tease apart the linker and compiler representations more, so we can put the check methods in the right places internally, and I just haven't gotten that done yet.

**Comment by eli-schwartz** [comment]
> This issue, once solved, can likely solve [#6486](https://github.com/mesonbuild/meson/issues/6486) and [#6487](https://github.com/mesonbuild/meson/issues/6487).

Not exactly, since the test still tries to assume it passes :) but either:
- on all platforms, it is correctly detected as working, and links properly, but running the resulting binaries only works if LD_PRELOAD is exclusively locked for use by asan
- on some platforms, the resulting error message is much better -- it fails at configure -- but the test still fails.

Re the run_unittests reproducer:

> From my understanding of the decorator around the tests that use sanitizers, they should be skipped when unsupported, but this doesn't happen.

The decorator checks if the compiler class (e.g. `icc`) doesn't recognize the option / doesn't list it as a setup option -- not if the compiler class does recognize it but needs to check whether the current build of the compiler is missing it.

Aside: this was incorrectly tagged as fixed by the wrong commit, but right PR. The fix (to check at setup time) landed in the preceding commit as e629d191b99119c360605b056b954fff0905c83f.
